crates/ruff_linter/src/rules/flake8_pyi/rules/generic_not_last_base_class.rs

```diff
@@ -0,0 +1,174 @@
+///     def push(self, item: T) -> None:
+///         self._items.append(item)
+///
+/// class MyMapping(
+///     Generic[K, V],
+///     Iterable[Tuple[K, V]],
+///     Container[Tuple[K, V]],
+/// ):
+///     ...
+/// ```
+///
+/// ```python
+/// class LinkedList(Sized, Generic[T]):
+///     def push(self, item: T) -> None:
+///         self._items.append(item)
+///
+/// class MyMapping(
+///     Iterable[Tuple[K, V]],
+///     Container[Tuple[K, V]],
+///     Generic[K, V],
+/// ):
+///     ...
+/// ```
+/// ## References
+/// - [`typing.Generic` documentation](https://docs.python.org/3/library/typing.html#typing.Generic)
+///
+/// [1]: https://github.com/python/cpython/issues/106102
+#[violation]
+pub struct GenericNotLastBaseClass;
+
+impl Violation for GenericNotLastBaseClass {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("`Generic[]` should always be the last base class")
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Move `Generic[]` to be the last base class".to_string())
+    }
+}
+
+/// PYI059
+pub(crate) fn generic_not_last_base_class(
+    checker: &mut Checker,
+    class_def: &StmtClassDef,
+    bases: Option<&Arguments>,
+) {
+    let Some(bases) = bases else {
+        return;
+    };
+
+    let semantic = checker.semantic();
+
+    let generic_base_indices: Vec<usize> = bases
+        .args
+        .iter()
+        .enumerate()
+        .filter_map(|(base_index, base)| {
+            if is_generic(base, semantic) {
+                return Some(base_index);
+            }
+            None
+        })
+        .collect();
+
+    if generic_base_indices.is_empty() {
+        return;
+    }
+
+    if generic_base_indices.len() == 1 {
+        let base_index = generic_base_indices[0];
+        if base_index == bases.args.len() - 1 {
+            // Don't raise issue for the last base.
+            return;
+        }
+
+        let mut diagnostic = Diagnostic::new(GenericNotLastBaseClass, class_def.identifier());
+        diagnostic.set_fix(generate_fix(bases, base_index, checker.locator()));
+        checker.diagnostics.push(diagnostic);
+    } else {
+        // No fix if multiple generics are seen in the class bases.
+        checker.diagnostics.push(Diagnostic::new(
+            GenericNotLastBaseClass,
+            class_def.identifier(),
+        ));
+    }
+}
+
+/// Return `true` if the given expression resolves to `typing.Generic[...]`.
+fn is_generic(expr: &Expr, semantic: &SemanticModel) -> bool {
+    if !semantic.seen_typing() {
+        return false;
+    }
+
+    let Expr::Subscript(ast::ExprSubscript { value, .. }) = expr else {
+        return false;
+    };
+
+    let qualified_name = semantic.resolve_qualified_name(value);
+    qualified_name.as_ref().is_some_and(|qualified_name| {
+        semantic.match_typing_qualified_name(qualified_name, "Generic")
+    })
```

Comment:
This can be written more simply as the following (which does exactly the same thing as your current logic under the hood):

```suggestion
    semantic.match_typing_expr(value, "Generic")
```